File: scripts/test_f32_sierpinski_cell.mjs
#!/usr/bin/env node
import assert from "node:assert/strict";
import { readFile } from "node:fs/promises";
import {
  F32_CELL_CLAIMS,
  F32_CELL_CONTRACT,
  MAX_F32_BUNDLE_CELLS,
  buildCellBundle,
  classifyWord,
  exactCellVertices,
  f32ToWord,
  tritsToWord,
  validateCell,
  verifyCellBundle,
  wordToCell,
  wordToTrits,
} from "../web/genomic-spectrum/f32-cell.js";

const profilePath = process.argv[2] ?? "conformance/f32_sierpinski_cell_v1.json";
const profile = JSON.parse(await readFile(profilePath, "utf8"));
assert.equal(profile.contract, F32_CELL_CONTRACT);
assert.deepEqual(profile.expected_claims, F32_CELL_CLAIMS);
const words = profile.words_hex.map((value) => Number.parseInt(value, 16) >>> 0);
const bundle = await buildCellBundle(words, profile.fields);
assert.equal(bundle.canonical_sha256, profile.expected_bundle_sha256);
assert.deepEqual(await verifyCellBundle(bundle), words);

for (const word of words) {
  const trits = wordToTrits(word);
  assert.equal(tritsToWord(trits), word);
  assert.equal(validateCell(wordToCell(word)), word);
}

let state = 0x5f3759df;
const randomWords = [];
for (let index = 0; index < 10_000; index += 1) {
  state = (Math.imul(state, 1664525) + 1013904223) >>> 0;
  randomWords.push(state);
}
assert.deepEqual(await verifyCellBundle(await buildCellBundle(randomWords)), randomWords);

const vertices = exactCellVertices(wordToTrits(0x5f3759df));
assert.equal(vertices.denominator, 2 ** 21);
for (const vertex of vertices.numerators) assert.equal(vertex.reduce((a, b) => a + b, 0), vertices.denominator);
assert.equal(classifyWord(0x7fc00001), "quiet-nan");
assert.equal(classifyWord(0x7fa00001), "signaling-nan");
assert.equal(f32ToWord(1.0), 0x3f800000);
assert.equal(f32ToWord(-0.0), 0x80000000);
assert.equal(f32ToWord(Math.PI), 0x40490fdb);

const tampered = structuredClone(wordToCell(0x3f800000));
tampered.claims.physical_storage_demonstrated = true;
assert.throws(() => validateCell(tampered), /claim boundary/);
const numericClaim = structuredClone(wordToCell(0x3f800000));
numericClaim.claims.physical_storage_demonstrated = 0;
assert.throws(() => validateCell(numericClaim), /claim boundary/);
assert.throws(() => wordToCell(0, { field: "chré" }), /printable ASCII/);
assert.throws(
  () => buildCellBundle(Array(MAX_F32_BUNDLE_CELLS + 1).fill(0)),
  new RegExp(String(MAX_F32_BUNDLE_CELLS)),
);

console.log(JSON.stringify({
  schema: "RSH-F32-SIERPINSKI-CELL-CROSS-RUNTIME-V1",
  status: "PASS",
  contract: F32_CELL_CONTRACT,
  cell_count: words.length,
  bundle_sha256: bundle.canonical_sha256,
  claims: F32_CELL_CLAIMS,
}, null, 2));

File: tests/test_f32_sierpinski_cell.py
# This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
# SPDX-License-Identifier: MPL-2.0
from __future__ import annotations

import copy
import json
import random
import unittest
from pathlib import Path

from scripts.f32_sierpinski_cell import (
    CLAIMS,
    DEPTH,
    build_bundle,
    classify_word,
    exact_cell_vertices,
    f32_to_word,
    trits_to_word,
    validate_cell,
    verify_bundle,
    verify_profile,
    word_to_cell,
    word_to_trits,
)

ROOT = Path(__file__).resolve().parents[1]
PROFILE = ROOT / "conformance" / "f32_sierpinski_cell_v1.json"


class F32SierpinskiCellTests(unittest.TestCase):
    def test_capacity_and_exact_round_trip(self):
        words = [
            0x00000000, 0x80000000, 0x00000001, 0x007FFFFF,
            0x00800000, 0x3F800000, 0xBF800000, 0x40490FDB,
            0x5F3759DF, 0x7F7FFFFF, 0x7F800000, 0xFF800000,
            0x7FC00001, 0x7FA00001, 0xFFFFFFFF,
        ]
        for word in words:
            trits = word_to_trits(word)
            self.assertEqual(len(trits), DEPTH)
            self.assertEqual(trits_to_word(trits), word)
            self.assertEqual(validate_cell(word_to_cell(word)), word)

    def test_random_word_round_trip(self):
        randomizer = random.Random(0x5F3759DF)
        words = [randomizer.getrandbits(32) for _ in range(10_000)]
        bundle = build_bundle(words)
        self.assertEqual(verify_bundle(bundle), words)

    def test_exact_cell_vertices_have_power_of_two_denominator(self):
        vertices, denominator = exact_cell_vertices(word_to_trits(0x5F3759DF))
        self.assertEqual(denominator, 1 << DEPTH)
        for vertex in vertices:
            self.assertEqual(sum(vertex), denominator)

    def test_ieee_classification_preserves_payload_classes(self):
        expected = {
            0x00000000: "zero",
            0x80000000: "zero",
            0x00000001: "subnormal",
            0x3F800000: "normal",
            0x7F800000: "infinity",
            0x7FC00001: "quiet-nan",
            0x7FA00001: "signaling-nan",
        }
        for word, classification in expected.items():
            self.assertEqual(classify_word(word), classification)

    def test_quake_magic_word_is_lineage_fixture_not_algorithm(self):
        cell = word_to_cell(0x5F3759DF, field="quake3-lineage-word")
        self.assertEqual(cell["word_hex"], "5f3759df")
        self.assertEqual(trits_to_word(cell["address_trits"]), 0x5F3759DF)
        self.assertFalse(cell["claims"]["compression_claim"])

    def test_tampering_is_rejected(self):
        cell = word_to_cell(0x3F800000)
        tampered = copy.deepcopy(cell)
        tampered["address_trits"] = "2" + tampered["address_trits"][1:]
        with self.assertRaisesRegex(ValueError, "outside|disagree"):
            validate_cell(tampered)
        tampered = copy.deepcopy(cell)
        tampered["cell_vertex_barycentric_numerators"][0][0] += 1
        with self.assertRaisesRegex(ValueError, "vertex evidence"):
            validate_cell(tampered)
        tampered = copy.deepcopy(cell)
        tampered["claims"]["physical_storage_demonstrated"] = True
        with self.assertRaisesRegex(ValueError, "claim boundary"):
            validate_cell(tampered)

    def test_numeric_projection_is_binary32(self):
        self.assertEqual(f32_to_word(1.0), 0x3F800000)
        self.assertEqual(f32_to_word(-0.0), 0x80000000)
        self.assertEqual(f32_to_word(3.141592653589793), 0x40490FDB)

    def test_profile(self):
        result = verify_profile(PROFILE)
        profile = json.loads(PROFILE.read_text(encoding="utf-8"))
        self.assertEqual(result["bundle_sha256"], profile["expected_bundle_sha256"])
        self.assertEqual(result["claims"], CLAIMS)


if __name__ == "__main__":
    unittest.main()
